[tutorials] Working with Lists tutorial (#730)
lit/lit.dev: packages/lit-dev-content/site/tutorials/content/working-with-lists/01.md

> If we're not working with arrays, we can also create our own array of

Reviewer: Hmm this set not being an array doesn't seem to be the problem in this step that you're tackling. Lit can still render a Set because it's an iterable e.g. just render `this.items` and see.

The problem being solved here is that you need to run .map on an iterable but .map is only available on Array not the generic iterable. Thus you can use `for...of`.

I'd also consider changing the title of this step to address the crux of lesson like `Using for Loops`

> If we're not working with arrays, we can also create our own array of

Reviewer: we

templates to render, for example, by using a `for...of` loop.

> ```ts

Reviewer: filename

const templates = [];
for (const item of this.items) {
  templates.push(html`<li>${item}</li>`);
}
```

It can then be placed inside the template returned from the render method.

> ```ts

Reviewer: filename

return html`
<ul>
  ${templates}
</ul>
`

> ```

Reviewer: perhaps lead into the next step a bit. something like "it works but is uggo code"

Sometimes you may want to render a list from another type of iterable,
such as a `Set`. Note the data in this example has been changed from an
array to a `Set`.

Lit provides a built-in `map` directive that will map over any iterable

> including `Map`s, `Set`s, and even generators, in addition to arrays.

Reviewer 2: Also kind of awkward to say, "a map directive that will map over ... maps"
When pluralizing a class name like this, I usually try to either make it generic ("maps, sets, and arrays"), add a noun like "objects": "`Map` and `Set` objects", or rephrase the sentence to use the singular. 

Maybe: "... a built-in `map` directive that works on any iterable: arrays, maps, sets, and even generators."

> When not working with an array as a source of data, an array of templates to
> render can be manually built, for example, by using a `for ... of` loop.

Reviewer 3: I didn't notice the example had changed from an array to a `Set` at first, so might be worth calling attention to that change first?
```suggestion
Sometimes you may want to render a list from something other than
an array, such as a `Set`. Note we've changed the example
data from an array to a `Set`.

You can manually manually build an array of templates from the `Set`
by using a `for ... of` loop, for example.
```

Reviewer 4: +1

Author: Fixed!

{% switchable-sample %}

```ts
// my-element.ts

> const templates = [];

Reviewer 3: More stylistic question for @e111077: I had my snippets include the indentation needed so that after copy/paste you didn't manually indent them to make it look ok. But it does make the snippets look kinda weird since they're floating right more. Maybe something to decide and normalize once all the tutorials are committed.

Reviewer: TBH I think it's fine removing the indents if it makes it look weird but i gotta say was nice for speed running your tutorial

for (const item of this.items) {
  templates.push(html`<li>${item}</li>`);
}
```

```js
// my-element.js
const templates = [];
for (const item of this.items) {
  templates.push(html`<li>${item}</li>`);
}
```

{% endswitchable-sample %}

> It can then be placed inside the template returned from the render method.

Reviewer 4: This might be too advanced but there's an opportunity here to mention (perhaps in an aside) that Lit templates can host any javascript expression, but if you're using statements like this, you can build up a result to later place in a template (statements can also be easily used by returning a result from a function... hopefully there's a link in the docs we could just call out to?).

Author: Added an aside stating any javascript expressions resulting in valid values for expressions with link to doc.

{% switchable-sample %}

```ts
// my-element.ts
return html`
  <ul>
    ${templates}
  </ul>
`
```

```js
// my-element.js
return html`
  <ul>
    ${templates}
  </ul>
`
```

{% endswitchable-sample %}

This is especially useful when you're not directly working with a single
iterable and are buliding up a list from multiple sources. However, if working

> with a single iterable, there is more declarative way of writing this shown in

Reviewer 5: Nit: "there is **a** more"

> Sometimes there may be a need to render a list from something other than an

Reviewer 6: I'd mention "iterable" in here, since that's what Lit supports, so like "render a list from another type of iterable, such as a `Set`"

> Sometimes there may be a need to render a list from something other than an

Reviewer 6: I think there's a question of ordering here... is .map(), imperative, map(), repeat(), the best order? map() is probably 99% of the time more readable than imperative, so maybe it should come first?

Author: That makes sense. I felt that starting with `map()` then reverting to imperative for the set would be silly, but perhaps the imperative step could be combined with the heterogeneous list idea below and to follow `map()`? What do you think about the placing of the events step? Before or after `repeat()`? I feel it could go either way, but leaning toward after.

> While Lit allows rendering values straight out of an iterable, it would be nice
> to have the ability to apply some transformation like adding markup for each
> item. Lit provides a built-in `map` directive that works on any iterable such as

Reviewer 6: ```suggestion
Lit supports rendering iterables of items directly in templates, but you often need to render a _template_ that uses the items, rather than the items directly. Lit provides a built-in `map` directive that lets you transform the items in an iterable, usually to create a template from them. The map directive works with any iterable such as...
```

arrays, sets, maps, and even generators. It returns an iterable containing the
result of calling the provided callback function on each item.

In this example, a custom element is presented with a property named `items`
that contains a set of strings.

Render each item in the set as a list item (`<li>`) containing that string.
Begin by importing the `map` directive.

{% switchable-sample %}

```ts
// my-element.ts
import {map} from 'lit/directives/map.js';
```

```js
// my-element.js
import {map} from 'lit/directives/map.js';
```

{% endswitchable-sample %}

> Use the directive in the rendered template, providing a callback function that

Reviewer 6: I don't know about "rendered template" - all templates are ultimately rendered, yeah? So just "template" or maybe "component's template" if you want something more?

returns a template wrapping each item with a list item (`<li></li>`) tag, and

> place it within the unordered list (`<ul></li>`) of the component's template.

Reviewer 6: ```suggestion
place it within the unordered list (`<ul></li>`) in the component's template.
```

Reviewer 6: You could also just refer to the `<ul>` tag rather than "the unordered list":
```suggestion
place it inside the `<ul>` tag in the template.
```

Reviewer 6: Oh, and `</li>` needs to be updated
```suggestion
place it within the unordered list (`<ul></ul>`) of the component's template.
```

Author: I like the shorter "`<ul>` tag".

{% switchable-sample %}

> ```ts

Reviewer 6: Do we need separate ts/js samples when they're exactly the same? @aomarks

Author: For most of these, I've added separate samples because of the first line comment that names the file and wanted the file extensions to match the selected language. I had a comment about adding the file name for where the code snippet should go, though arguably since there's only ever 1 file you're editing for this tutorial it could be omitted.

Reviewer: I'd say it's okay to omit the switchable-sample macro in these cases

Author: Should I do
```
// my-element.(js|ts)
```
for the first line comment in the samples then, or omit the filename comment all-together for this tutorial since the reader is only ever working on the one file?

Maybe this is another thing that we'd want a consistency pass across all tutorials as well. @arthurevans what do you think?

Reviewer 6: Oh... that's enough of a difference to keep the two then. I missed that and thought they were identical.
